tx-indexer/src/collector.rs

```diff
@@ -292,81 +297,83 @@ async fn save_transaction_details(
         }
     };
     let transaction_hash = transaction_details.transaction.hash.to_string();
-    let signer_id = transaction_details.transaction.signer_id.to_string();
-    let tx_bytes = &borsh::to_vec(&transaction_details).expect("Failed to serialize transaction");
+    let tx_bytes = borsh::to_vec(&transaction_details).expect("Failed to serialize transaction");
 
-    match db_manager
-        .add_transaction(
-            &transaction_hash,
-            tx_bytes.clone(),
-            tx_details.block_height,
-            &signer_id,
-        )
-        .await
-    {
-        Ok(_) => {
-            // Validate that the transaction is saved correctly
-            // Before validating, we need to wait a little bit to make sure that the transaction is saved
-            tokio::time::sleep(std::time::Duration::from_millis(150)).await;
-            let mut attempts = 1;
-            let mut tx_bytes = tx_bytes.clone();
-            while let Err(err) = db_manager
-                .validate_saved_transaction_deserializable(&transaction_hash, &tx_bytes)
-                .await
-            {
-                tracing::warn!(
+    // We faced the issue when the transaction was saved to the storage but later
+    // was failing to deserialize. To avoid this issue and monitor the situation
+    // without runing the user experience, we will try to save the transaction
+    // to the storage and validate that it is saved correctly for 3 attempts
+    // before we throw an error.
+    let mut save_attempts = 0;
+    'retry: loop {
+        save_attempts += 1;
+        if save_attempts > 3 {
+            tracing::error!(
+                target: crate::INDEXER,
+                "Failed to save transaction {} after 3 attempts",
+                transaction_hash
+            );
+            break false;
+        }
+        match tx_details_storage
+            .store(&transaction_hash, tx_bytes.clone())
+            .await
+        {
+            Ok(_) => {
+                // At this moment transaction seems to be stored, and we want to validate the correctness of the stored data
+                // To validate we will try to retrieve the transaction from the storage and validate that it is deserializable
+                // If the transaction is not deserializable, we will try to save it again
+                'validator: loop {
+                    tokio::time::sleep(std::time::Duration::from_millis(150)).await;
+                    let Ok(tx_details_bytes_from_storage) =
+                        tx_details_storage.retrieve(&transaction_hash).await
+                    else {
+                        tracing::error!(
+                            target: crate::INDEXER,
+                            "Failed to retrieve transaction {} from storage",
+                            transaction_hash,
+                        );
+                        continue 'validator;
```

Comment:
I think we need to add a separate counter for this loop in case we hit infinity for some unknown reason, then we can get out of it. What do you think?